lib/src/navbar/navbar.component.ts

```diff
@@ -1,55 +1,126 @@
-import { Component, OnInit, Input, QueryList, ContentChildren } from '@angular/core';
-import { NavbarMobileMenuComponent } from './navbar-mobile-menu/navbar-mobile-menu.component';
-
-@Component({
-    selector: 'hc-navbar',
-    templateUrl: './navbar.component.html',
-    styleUrls: ['./navbar.component.scss']
-})
-export class NavbarComponent implements OnInit {
-
-    @Input() user: string = '';
-
-    @Input() appIcon: string = '';
-
-    @Input() brandIcon: string = '';
-
-    @Input() homeUri: string = '';
-
-    @Input() fixedTop: boolean = false;
-
-    @ContentChildren(NavbarMobileMenuComponent) mobileMenu: QueryList <NavbarMobileMenuComponent>;
-
-    menuOpen: boolean = false;
-
-    constructor() {
-    }
-
-    ngOnInit() {
-    }
-
-    toggleMobileMenu() {
-        if ( this.mobileMenu.first ) {
-            if ( this.menuOpen ) {
-                this.mobileMenu.first.hide();
-                this.menuOpen = false;
-            } else {
-                this.mobileMenu.first.show();
-                this.menuOpen = true;
-            }
-        }
-    }
-
-    menuClick( event: any ) {
-        let clickTarget: string = event.target.classList[0];
-
-        // Verify that the click in the mobile menu came from a navigation item
-        if ( clickTarget.indexOf('menu-dropdown') === -1 ) {
-            this.toggleMobileMenu();
-        }
-    }
-
-    get mobileMenuIcon(): string {
-        return this.menuOpen ? 'fa-times' : 'fa-bars';
-    }
-}
+import { Component, Input, QueryList, ContentChildren, ElementRef, HostListener,
+         AfterViewInit, ChangeDetectorRef, ViewChild } from '@angular/core';
+import { NavbarMobileMenuComponent } from './navbar-mobile-menu/navbar-mobile-menu.component';
+import { NavbarLinkComponent } from './navbar-link/navbar-link.component';
+import { PopoverContentComponent } from '../popover/popoverContent.component';
+
+@Component({
+    selector: 'hc-navbar',
+    templateUrl: './navbar.component.html',
+    styleUrls: ['./navbar.component.scss']
+})
+export class NavbarComponent implements AfterViewInit {
+
+    @Input() user: string = '';
+    @Input() appIcon: string = '';
+    @Input() brandIcon: string = '';
+    @Input() homeUri: string = '';
+    @Input() fixedTop: boolean = false;
+
+    @ContentChildren(NavbarMobileMenuComponent) mobileMenu: QueryList <NavbarMobileMenuComponent>;
+    @ContentChildren(NavbarLinkComponent) navLinks: QueryList <NavbarLinkComponent>;
+    @ViewChild(PopoverContentComponent) popper: PopoverContentComponent;
+
+    menuOpen: boolean = false;
+    linkWidths: Array<number> = [];
+    linksMax: number = 0;
+    collapse: boolean = false;
+    logoCondense: boolean = false;
+    logoWidth: number = 0;
+    moreList: Array<Object> = [];
+
+    @HostListener('window:resize') navResize() {
+        this.popper.hide();
```

Comment:
This is some awesome & complex logic. Couple thoughts on it that don't necessarily need to be addressed now:

- Most components don't have a whole lot of real unit tests, but this type of code is prime for testing. Maybe not now, but at some point it would be good to have some solid tests here.

- This function might be a little more digestible if broken up into smaller functions. Func names nicely document/explain what's going on, and are more easily unit testable. I could see there being at least a couple private functions that navResize() calls... maybe `condenseLogoIfNeeded()` and `collapseNavLinksIfNeeded()`. Also looks like lines 54-56 and 60-62 are almost identical, could be a simple private function there, too.

Reply:
Wow, this is awesome. This was a lot of work